cpp/ql/src/experimental/Security/CWE/CWE-416/IteratorToExpiredContainer.ql

```diff
@@ -86,12 +70,39 @@ predicate destroyedToBeginSink(DataFlow::Node sink, FunctionCall fc) {
 }
 
 /**
- * Flow from any destroyed object to the qualifier of a `begin` or `end` call
+ * A configuration to track flow from a temporary variable to the qualifier of
+ * a destructor call, and subsequently to a qualifier of a call to `begin` or
+ * `end`.
  */
-module DestroyedToBeginConfig implements DataFlow::ConfigSig {
-  predicate isSource(DataFlow::Node source) { source = getADestroyedNode() }
+module Config implements DataFlow::StateConfigSig {
+  newtype FlowState =
+    additional TempToDestructor() or
+    additional DestroyedToBegin(DataFlow::Node n) {
+      exists(DataFlow::Node thisOperand |
+        tempToDestructorSink(thisOperand, _) and
+        n = getADestroyedNode(thisOperand)
+      )
+    }
+
+  predicate isSource(DataFlow::Node source, FlowState state) {
+    source.asInstruction().(VariableAddressInstruction).getIRVariable() instanceof IRTempVariable and
+    state = TempToDestructor()
+  }
+
+  predicate isAdditionalFlowStep(
+    DataFlow::Node node1, FlowState state1, DataFlow::Node node2, FlowState state2
+  ) {
+    tempToDestructorSink(node1, _) and
+    state1 = TempToDestructor() and
+    state2 = DestroyedToBegin(node2) and
+    node2 = getADestroyedNode(node1)
+  }
```

Comment:
So what exactly is different now compared to when we had two sequential dataflows?  The source of the second flow (`DestroyedToBeginFlow`) still had to be a reached sink of the first (`TempToDestructorFlow`) flow.  I can see that we now know the correct overall source ... but that isn't reported by the query anyway.

Reply:
Consider this hypothetical example:
```cpp
int id(int x) {
  return x;
}

void f1() {
  int x2 = id(source());
  safe(x2);
}

void f2() {
  int x1 = id(0);
  sink(x1);
}
```
And consider a QL query that has two configurations:

- One configuration (`C1`) from `source()` to `x` in `return x` (in `id`), and
- One configuration (`C2`) starting at any sink from `C1` to the argument of `sink`.

`C1` finds a flow path from `f1` to `id`, and `C2` finds a flow path from `id` to `f2`. So if we simply use the result of `C2` we get a FP saying that there's flow from `source()` to `sink(x1)`. However, if we combine these into a single configuration then the dataflow library won't lose track of the call context, and cause flow to enter `id` from `f1` and exit `id` to `f2`.

Does that make sense?

Reply:
Ah yes, that makes sense and fits with what's going on in the full test case.  Thanks!